### PR TITLE
Use schema registry in alerting services

### PR DESCRIPTION
**Pull request by rish691**
https://github.com/hypertrace/hypertrace-alert-engine/issues/77

### CODE REVIEW

File: metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaConfigReader.java [hunk 1]
```diff
@@ -1,24 +1,39 @@
...
-  public String getTopicName() {
-    return kafkaQueueConfig.getString(TOPIC_NAME_CONFIG);
+  public String getConsumerTopicName() {
+    return kafkaQueueConfig.getString(
+        new StringJoiner(".").add(CONSUMER_CONFIG).add(TOPIC_NAME_CONFIG).toString());
+  }
+
+  public String getProducerTopicName() {
+    return kafkaQueueConfig.getString(
+        new StringJoiner(".").add(CONSUMER_CONFIG).add(TOPIC_NAME_CONFIG).toString());
```
Comment (kotharironak) [inline metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaConfigReader.java:36]: Should this be `PRODUCER_CONFIG`?

File: metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaAlertTaskConsumer.java [hunk 2]
```diff
@@ -61,12 +60,7 @@ public void close() {
     baseProperties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
     baseProperties.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, "1000");
     baseProperties.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, "30000");
-    baseProperties.put(
-        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
-        "org.apache.kafka.common.serialization.StringDeserializer");
-    baseProperties.put(
-        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
-        "org.apache.kafka.common.serialization.ByteBufferDeserializer");
+    baseProperties.putAll(kafkaConfigReader.getConsumerSerdeProperties());
```
Comment (kotharironak) [inline metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaAlertTaskConsumer.java:63]: Just a note, when you work on deployment, for this consumer component, we will need a higher value for `heartbeat.interval.ms` and `session.timeout.ms`.

File: metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaConfigReader.java [hunk 3]
```diff
@@ -33,22 +48,55 @@ public String getBootstrapServer() {
     return mergeProperties(baseProps, getFlatMapConfig(kafkaQueueConfig, CONSUMER_CONFIG));
   }
 
+  public Map<String, Object> getProducerSerdeProperties() {
+    String schemaRegistryUrl =
+        new StringJoiner(".").add(CONSUMER_CONFIG).add(SCHEMA_REGISTRY_URL).toString();
+    Map<String, Object> map = new HashMap<>();
+    if (kafkaQueueConfig.hasPath(schemaRegistryUrl)) {
```
Comment (kotharironak) [inline metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaConfigReader.java:55]: we can put all these additional properties in application.conf file in PRODUCER_CONFIG section eg. `schema.registry.url`, and serde properties
we just have to put the default value in the corresponding class's base properties. So, would suggest, add below to base properties:

```
ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, AvroSerializer.class.getName()

ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, AvroSerializer.class.getName()
```
And, update the application.conf file with 
```
schema.registry.url = 
key.serializer= 
value.serializer
```
The method `getProducerConfig` merges application.conf properties with baseProperties

File: notification-service/src/main/java/org/hypertrace/alert/engine/notification/service/NotificationEventProcessorService.java [hunk 4]
```diff
@@ -48,15 +46,10 @@ protected void doInit() {}
...
-      try {
-        ConsumerRecords<String, ByteBuffer> records =
-            consumer.poll(Duration.ofMillis(CONSUMER_POLL_TIMEOUT_MS));
-        for (ConsumerRecord<String, ByteBuffer> record : records) {
-          notificationEventProcessor.process(
-              MetricAnomalyNotificationEvent.fromByteBuffer(record.value()));
-        }
-      } catch (IOException e) {
-        LOGGER.error("Exception processing record", e);
+      ConsumerRecords<String, MetricAnomalyNotificationEvent> records =
```
Comment (kotharironak) [inline notification-service/src/main/java/org/hypertrace/alert/engine/notification/service/NotificationEventProcessorService.java:49]: How would the serde issue be handled? In such cases, service will die, right?

**Review by kotharironak — COMMENTED** [review]
overall, lgtm

File: notification-service/src/main/java/org/hypertrace/alert/engine/notification/service/NotificationEventProcessorService.java (hunk 4 above)
Comment (rish691) [reply notification-service/src/main/java/org/hypertrace/alert/engine/notification/service/NotificationEventProcessorService.java:49]: What do you mean by Serde issue?

File: metric-anomaly-data-model/build.gradle.kts [hunk 5]
```diff
@@ -9,10 +9,12 @@ dependencies {
     api("io.grpc:grpc-protobuf:1.37.0")
     api("io.grpc:grpc-stub:1.37.0")
     api("javax.annotation:javax.annotation-api:1.3.2")
-    api("org.apache.avro:avro:1.10.2")
     implementation("com.typesafe:config:1.4.1")
     implementation("org.apache.kafka:kafka-clients:2.6.0")
     implementation("org.hypertrace.core.documentstore:document-store:0.5.7")
+    api("org.apache.avro:avro:1.10.2")
```
[the remark this answers is not in the document]
Comment (rish691) [reply metric-anomaly-data-model/build.gradle.kts]: one is from hypertrace and other is from confluent

File: metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaConfigReader.java (hunk 3 above)
Comment (rish691) [reply metric-anomaly-data-model/src/main/java/org/hypertrace/alert/engine/metric/anomaly/datamodel/queue/KafkaConfigReader.java]: Makes sense, updated

File: metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/NotificationEventProducer.java [hunk 6]
```diff
@@ -1,62 +1,30 @@
...
 
   public void publish(MetricAnomalyNotificationEvent notificationEvent) {
-    try {
-      producer.send(
-          new ProducerRecord<String, ByteBuffer>(
-              properties.getProperty("topic"), null, notificationEvent.toByteBuffer()));
-    } catch (IOException e) {
-      LOGGER.error("Exception producing messages", e);
-    }
+    producer.send(new ProducerRecord<>(properties.getProperty("topic"), null, notificationEvent));
```
Comment (kotharironak) [inline metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/NotificationEventProducer.java:24]: As a follow-up, we may have to hook ErrorHandler in properties to handle serde errors if any.

**Review by kotharironak — APPROVED** [review]
overall lgtm

**Pull request merged**